Openstack dynamic inventory: respect OS_CLOUD env var

alvaroaleman opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

Openstack dynamic inventory
##### ANSIBLE VERSION

```
devel
```
##### SUMMARY

RIght now the Openstack inventory script always requests -all- defined clouds. This is an extremely expensive operation. This change makes it possible to use the OS_CLOUD environment variable to specify a cloud the same way the openstack command uses it:

``` shell
openstack --help|grep -B1 OS_CLOUD
  --os-cloud <cloud-config-name>
                        Cloud name in clouds.yaml (Env: OS_CLOUD)
```

alvaroaleman committed:
OS dynamic inventory: respect OS_CLOUD env var

alvaroaleman commented:
CC @emonty

rcarrillocruz reviewed (commented):
I like this, actually asked Monty about something similar a while ago.

Do you think you could add an option to read a cloud from CLI args too?
If both CLI and envvars are present, CLI prevails.

halberom commented:
Hello,  I've had a quick play with something very similar to this.  Several points
- adding cli option can be as easy as adding the following diff
```
197,198d196
<     parser.add_argument('--cloud', default=os.environ.get('OS_CLOUD'),
<                         help='Cloud name (default: None')
223d220
<             cloud=args.cloud,
```
- my main concern however is that if cache expiration is defined, you must _always_ refresh the cache when changing the cloud at present.  What would be nicer is if the cache was per cloud (similar to how the aws ec2.py does it - https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L394-L395).

alvaroaleman commented:
@halberom can you push that change somewhere so I can cherry-pick it into this PR?

Commit:
support OS_CLOUD in openstack dynamic inventory

allow a user to limit to a single cloud in clouds.yaml, and support
per-cloud caching.

halberom commented:
here you go https://github.com/halberom/ansible/blob/feature/openstack_os_cloud/contrib/inventory/openstack.py

alvaroaleman commented:
Thanks @halberom!

samdoran commented:
@rcarrillocruz Can you look at this again and see if it's ok to merge?

emonty reviewed (approved):
shipit

abadger commented:
Merged to devel for 2.4.0 release
